xdsmatcher/README.md

```diff
@@ -0,0 +1,109 @@
+# xDS Matching API Evaluation Framework
+
+This library provides a way to evaluate the
+[xDS matching API](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/matching/matching_api)
+using the same logic that's implemented within the Envoy proxy. The intended use case right now is 
+to support unit testing of generated configuration, but this may be extended to allow for the usage
+of this matching API in other contexts.
+
+## Example Usage
+
+To use, first register the input and action extensions that will be used:
+
+```golang
+func init() {
+	// Register the input type with the register, using the type URL of the describing protobuf.
+	registry.InputExtensions["type.googleapis.com/xdsmatcher.test.proto.FooInput"] = fooInputFactory
+}
+
+// The implementation of the input.
+type fooInput struct {
+}
+
+// The factory method of the input, this converts the describing protobuf into the input object.
+func fooInputFactory(m proto.Message) (interface{}, error) {
+	_, ok := m.(*pbtest.FooInput)
+	if !ok {
+		return nil, errors.New("unexpected proto")
+	}
+
+	return fooInput{}, nil
+}
+
+// Implement types.DataInput for the input type.
```

Comment:
Just so I understand correctly, is the `InputType` the registered type URL in the `init()` function? I'd assume so since we are attempting a cast below that never fails

Reply:
To make things super generic, the entire tree acts against a specific InputType (simple case here is just a string, but can be a complex data structure). Once the tree is built, you pass a value of type InputType to it, which is then passed to all DataInputs (which tell us how to extract a specific value out of the complex data type).

So the InputType is never registered, but you must be defining DataInputs that all support the specific type that you provide to the matcher on evaluation time